src/library/scala/collection/IterableOnce.scala

```diff
@@ -721,7 +721,7 @@ trait IterableOnceOps[+A, +CC[_], +C] extends Any { this: IterableOnce[A] =>
   def min[B >: A](implicit ord: Ordering[B]): A = {
     if (isEmpty)
       throw new UnsupportedOperationException("empty.min")
-    reduceLeft((x, y) => if (ord.lteq(x, y)) x else y)
+    reduceLeft(ord.min).asInstanceOf[A] // need asInstanceOf because ord.min returns B
```

Comment:
This is technically unsound because the `min` of the ordering could return a value that isn't either of the inputs.

Reply:
Yeah, I thought of that. Not sure what a good solution is though. As you noted, the current way is broken too.

Reply:
Yes, I think it should if it's not too much of a hassle.

Reply:
This seems like a weird PR to make that change, but I can make that change if you want (I'll put it in a separate commit)

Reply:
This is broken as follows:

```
[error] def min[U <: Double](x: U,y: U): U
[error]     override def min(x: Double, y: Double): Double = math.min(x, y)
```

so I would need to implement it as:

```scala
trait DoubleOrdering extends Ordering[Double] {
  // ...
  override def min[U <: Double](x: U, y: U) = math.min(x, y) // can't call math.min on type U
}
```

Reply:
We could use the same encoding in `math.min` but I think it would complicate things unnecessarily. There's no direct benefit for a primitive type. Maybe in Dotty it can be expressed even more precisely with a union type (`x.type | y.type`) which could be useful even in `math.min`.

For now, if there are no major incompatibilities outside of the primitive `Ordering` implementations, I'd go with the precise type in `Ordering.min` (and `max` obviously) and live with the casts in the primitive implementations.

Reply:
Only the final cast should be necessary

Reply:
You're right, my mistake. Not sure why I thought those casts were needed